src/map/pc_groups.cpp

```diff
@@ -3,462 +3,356 @@
+
+						inherited = true;
+						break;
+					}
+				}
+
+				if( inherited ){
+					it = entry.second.erase( it );
 					continue;
-				ARR_FIND(0, ARRAYLENGTH(pc_g_permission_name), j, strcmp(pc_g_permission_name[j].name, name) == 0);
-				group_settings->e_permissions |= pc_g_permission_name[j].permission;
+				}else if( !found ){
+					ShowError( "Inherited group \"%s\" for group id %u does not exist.\n", otherName.c_str(), group->id );
+					it = entry.second.erase( it );
+					continue;
+				}else{
+					it++;
+				}
 			}
 		}
-		dbi_destroy(iter);
+
+		if( this->inheritance.empty() ){
+			break;
+		}
 	}
 
-	ShowStatus("Done reading '" CL_WHITE "%d" CL_RESET "' groups in '" CL_WHITE "%s" CL_RESET "'.\n", group_count, config_filename);
+	if( i == MAX_CYCLES && !this->inheritance.empty() ){
+		ShowError( "Could not process inheritance rules, check your config for cycles...\n" );
+	}
 
-	
-	if( ( pc_group_max = group_count ) ) {
-		DBIterator *iter = db_iterator(pc_group_db);
-		GroupSettings *group_settings = NULL;
-		int* group_ids = (int*)aMalloc( pc_group_max * sizeof(int) );
-		int i = 0;
-		for (group_settings = (GroupSettings *)dbi_first(iter); dbi_exists(iter); group_settings = (GroupSettings *)dbi_next(iter)) {
-			group_ids[i++] = group_settings->id;
-		}
-		
-		atcommand_db_load_groups(group_ids);
-		
-		aFree(group_ids);
-		
-		dbi_destroy(iter);
+	// Not needed anymore
+	this->inheritance.clear();
+
+	uint32 index = 0;
+	for( auto& it : *this ){
+		it.second->index = index++;
 	}
-}
 
-/**
- * Removes group configuration from memory.
- * @private
- */
-static void destroy_config(void)
-{
-	config_destroy(&pc_group_config);
+	// Initialize command cache
+	atcommand_db_load_groups();
 }
 
-/**
- * In group configuration file, setting for each command is either
- * <commandname> : <bool> (only atcommand), or
- * <commandname> : [ <bool>, <bool> ] ([ atcommand, charcommand ])
- * Maps AtCommandType enums to indexes of <commandname> value array,
- * COMMAND_ATCOMMAND (1) being index 0, COMMAND_CHARCOMMAND (2) being index 1.
- * @private
- */
-static inline int AtCommandType2idx(AtCommandType type) { return (type-1); }
+PlayerGroupDatabase player_group_db;
 
 /**
  * Checks if player group can use @/#command
- * @param group_id ID of the group
  * @param command Command name without @/# and params
  * @param type enum AtCommanndType { COMMAND_ATCOMMAND = 1, COMMAND_CHARCOMMAND = 2 }
  */
-bool pc_group_can_use_command(int group_id, const char *command, AtCommandType type)
-{
-	int result = 0;
-	config_setting_t *commands = NULL;
-	GroupSettings *group = NULL;
-
-	if (pc_group_has_permission(group_id, PC_PERM_USE_ALL_COMMANDS))
+bool s_player_group::can_use_command( const std::string& command, AtCommandType type ){
+	if( this->has_permission( PC_PERM_USE_ALL_COMMANDS ) ){
 		return true;
+	}
 
-	if ((group = id2group(group_id)) == NULL)
-		return false;
+	std::unordered_map<std::string, bool>* map;
```

Comment:
We could possibly speed this up with `std::vector<std::string>`.

Reply:
This is only used during load time.
Do you still think this is necessary?

Reply:
Done!